### ISSUE TITLE
Test all Intel CPUs in Travis CI via SDE

### ISSUE THREAD

**Issue by jeffhammond** [flame/blis#122]
@devinamatthews You will probably like this...

Intel [SDE](http://www.intel.com/software/sde) supports no less than the following list of architectures:
```
     -quark              Set chip-check and CPUID for Intel(R) Quark
     -p4                 Set chip-check and CPUID for Pentium4
     -p4p                Set chip-check and CPUID for Pentium4 Prescott
     -mrm                Set chip-check and CPUID for Merom
     -pnr                Set chip-check and CPUID for Penryn
     -nhm                Set chip-check and CPUID for Nehalem
     -wsm                Set chip-check and CPUID for Westmere
     -snb                Set chip-check and CPUID for Sandy Bridge
     -ivb                Set chip-check and CPUID for Ivy Bridge
     -hsw                Set chip-check and CPUID for Haswell
     -bdw                Set chip-check and CPUID for Broadwell
     -skx                Set chip-check and CPUID for Skylake Server
     -skl                Set chip-check and CPUID for Skylake Client
     -cnl                Set chip-check and CPUID for Cannonlake
     -knl                Set chip-check and CPUID for Knights Landing
     -slt                Set chip-check and CPUID for Saltwell
     -slm                Set chip-check and CPUID for Silvermont
     -glm                Set chip-check and CPUID for Goldmont
```

This means that one can run native binaries for these processors on other processors, e.g. whatever Travis CI happens to be using.

I use it on a daily basis to emulate KNL on HSW.  It will allow us to set `RUN_TEST=1` for SNB, HSW, KNL, etc.

I am working on https://github.com/jeffhammond/blis/tree/travis-ci-sde-knl-testing.

**Comment by devinamatthews** [comment]
Yes, I've used SDE for testing KNL and that other one. My concern would be build time of course as it is quite slow. Maybe only testing a subset of operations?

**Comment by jeffhammond** [comment]
@devinamatthews Compilation is unaffected, because cross-compiler works fine.  We should probably have an option to test only 100x100 but I was running KNL binaries on HSW yesterday and it seemed pretty fast.  I'll time it today.  The only unknown is whether something SDE uses is perturbed by the AWS environment.  But I can time SDE inside of a Debian VM locally if necessary.

**Comment by devinamatthews** [comment]
I tend to get in the O(0.1 GFLOP) range, which takes a while with all 
the tests that are run by default. But the proof is in the pudding.

**Comment by jeffhammond** [comment]
@devinamatthews Yeah, I killed the SDE timing I was running after 5+ hours so yeah, my memory is less reliable than DRAM in the vicinity of a brick of plutonium.

However, the bigger issue is that Travis CI does not allow me to run a 32-bit binary ([details](https://github.com/travis-ci/travis-ci/issues/986)) without forcing the entire BLIS project to enable sudo, which has a well-known performance impact ([details](https://blog.travis-ci.com/2014-12-17-faster-builds-with-container-based-infrastructure/)).

Hopefully I can get a 64-bit binary from the SDE folks.  If not, we will have to discuss whether KNL testing is worth reverting to slower VM builds instead of containers.

**Comment by jeffhammond** [comment]
The good news is the 32-bit binary issue was because **I AM A COMPLETE MORON**.  So yeah, I think I'll have this working tonight.

**Comment by fgvanzee** [comment]
Man, I *knew* I should have upgraded from the standard plutonium when my house was being built. :)

Thanks for your efforts, Jeff.

**Comment by devinamatthews** [comment]
Just a quick update: I just hacked up a custom CPUID definition file for SKX with a proper brand string so that 2 VPUs are detected, and it works (in TBLIS, but should work the same in BLIS). So, with a full set of CPUID files (which are pretty easy to generate/change) we can test at least the level 1v, 1f, and 3m ukernels for all non-FMA4 architectures on Travis. This is all Intel and AMD architectures except Bulldozer I think.

**Comment by fgvanzee** [comment]
Thanks for the update, Devin. Being able to test all architectures (except poor Bulldozer) in Travis would be quite appealing. Please point me to some code in TBLIS when you're done and settled so I can look into duplicating the setup in BLIS. (Or you can do a PR :) but I know you're busy these days.)

**Comment by devinamatthews** [comment]
I'll make up CPUID files for all archs and make a PR. Jeff's PR (#124) has the Travis-side stuff. QEMU is the other option -- I've tested it for ARM32/64 and I think it should do Bulldozer too. It doesn't do AVX512 so we'll need SDE in any case.

**Comment by devinamatthews** [comment]
@jeffhammond problem: the CPUID definition files in SDE have a restrictive copyright notice, do we need to clean-room them to distribute with BLIS?

**Comment by devinamatthews** [comment]
NVM: http://instlatx64.atw.hu/ has SKX dumps now.

**Comment by jeffhammond** [comment]
@devinamatthews Send me email at my Intel address to discuss.  I will try to figure it out.

**Comment by devinamatthews** [comment]
This is working in ab70a87, so I am going to close.

**Comment by wolfv** [comment]
Hey guys,

just a quick question as you might be able to help me out: I am trying to get SDE working on Travis to test our SIMD library (https://github.com/QuantStack/xsimd) as well but have a problem with the following error popping up. Tried to look through your code but I can't really find you configuring SDE much. Any chance you've seen this?

Btw. looking into BLIS as something to potentially use in https://github.com/QuantStack/xtensor ! :)


```
/home/travis/.travis/job_stages: line 57:  2697 Killed                  ../intel-sde-downloader/sde-external-8.16.0-2018-01-30-lin/sde64 -cnl -- ./test_xsimd
E: Attach to pid 2697 failed. 
E:   The Operating System configuration prevents Pin from using the default (parent) injection mode.
E:   To resolve this, either execute the following (as root):
E:   $ echo 0 > /proc/sys/kernel/yama/ptrace_scope
E:   Or use the "-injection child" option.
E:   For more information, regarding child injection, see Injection section in the Pin User Manual.
E: 
```

**Comment by devinamatthews** [comment]
@wolfv I think I managed to work around this with `sudo: required` in the travis.yml file. Thanks for your interest in BLIS and if you are working with tensors then definitely give [TBLIS](https://github.com/devinamatthews/tblis) a try, as it has native tensor contraction etc. (I'm also work on some Tucker/CP factorization stuff currently).

**Comment by wolfv** [comment]
Ok, actually I also finally got it to work with sudo required and some echo 0 > proc/yama... Thanks for the help! 

It's unfortunate that TBLIS has so little documentation. I couldn't find an example of the C++ interface. Do you have a generalized, numpy-style dot in there? That would be very interesting. Maybe you want to come over to gitter.im/QuantStack/Lobby for a quick chat?

**Comment by jeffhammond** [comment]
@wolfv SDE ships as a ready-to-go binary that you unpack and run.  Feel free to send me an email if you need help using SDE.
